### PR TITLE
Personalize Indigo Gift Voucher with Invalid promo code

### PR DESCRIPTION
**Pull request by mshahrukh2**
Personalize Indigo Gift Voucher with Invalid promo code and validating  payment amount is unchanged

### CODE REVIEW

File: src/test/resources/reportportal.properties [hunk 1]
```diff
@@ -1,11 +1,6 @@
...
-rp.mode=DEFAULT
-rp.skipped.issue=true
-rp.batch.size.logs=20
\ No newline at end of file
+ReportPortal configuration:
+rp.endpoint = http://reportportal.jio.com:8080
+rp.uuid = f8379b19-8b89-493e-b5bd-6ffc33096850
+rp.launch = teswiz_TEST_EXAMPLE
+rp.project = teswiz
+rp.enable = true
```
Comment (anandbagmar) [inline src/test/resources/reportportal.properties:6]: other properties will be important to have

File: src/test/resources/environments.json [hunk 2]
```diff
@@ -4,6 +4,7 @@
     "BASE_URL": "http://the-internet.herokuapp.com/",
     "IMAGES_BASE_URL": "http://images.google.com/",
     "BING_BASE_URL": "http://bing.com/images",
-    "JIOMEET_BASE_URL": "https://jiomeetpro.jio.com/home"
+    "JIOMEET_BASE_URL": "https://jiomeetpro.jio.com/home",
+    "Indigo_Base_URL": "https://www.goindigo.in/"
```
Comment (anandbagmar) [inline src/test/resources/environments.json:8]: follow similar naming convention

File: src/test/java/com/znsio/sample/e2e/steps/IndigoGiftVoucherSteps.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    @When("User applies invalid promocode")
+    public void userAppliesInvalidPromocode() {
+    new VoucherBL().applyPromoCode();
+    }
+
+    @Then("User should be able to proceed with orignal amount.")
+    public void userShouldBeAbleToProceedWithOrignalAmount() {
+        new VoucherBL().validatePurchase();
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/steps/IndigoGiftVoucherSteps.java:42]: add loggers

File: src/test/java/com/znsio/sample/e2e/screen/web/indigo/PromoScreenWeb.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    private final Visual visually;
+    private final WebDriver innerDriver;
+    private final TestExecutionContext context;
+    private static final String SCREEN_NAME = GiftVoucherScreenWeb.class.getSimpleName();
+    private static final Logger LOGGER = Logger.getLogger(SCREEN_NAME);
+
+    private final By byPromoCodexpath = By.xpath(("//*[@id=\"PromoCode\"]"));
+    private final By byApplyPromoCodeId = By.id("btnApplyPromoCode");
+    private final By byPromoCodeMessageXpath = By.xpath("//input[@id=\"PromoCode\"]//following-sibling::div");
+    private final By byTotalAmountxpath = By.xpath("(//*[@id=\"lblTotal\"])[1]");
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/web/indigo/PromoScreenWeb.java:23]: locator is bad

File: src/test/java/com/znsio/sample/e2e/screen/web/indigo/PromoScreenWeb.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    private final Driver driver;
+    private final Visual visually;
+    private final WebDriver innerDriver;
+    private final TestExecutionContext context;
+    private static final String SCREEN_NAME = GiftVoucherScreenWeb.class.getSimpleName();
+    private static final Logger LOGGER = Logger.getLogger(SCREEN_NAME);
+
+    private final By byPromoCodexpath = By.xpath(("//*[@id=\"PromoCode\"]"));
+    private final By byApplyPromoCodeId = By.id("btnApplyPromoCode");
+    private final By byPromoCodeMessageXpath = By.xpath("//input[@id=\"PromoCode\"]//following-sibling::div");
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/web/indigo/PromoScreenWeb.java:22]: locators should be private static final

File: src/test/java/com/znsio/sample/e2e/screen/web/indigo/PromoScreenWeb.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+public class PromoScreenWeb extends PromoScreen {
+    private final Driver driver;
+    private final Visual visually;
+    private final WebDriver innerDriver;
+    private final TestExecutionContext context;
+    private static final String SCREEN_NAME = GiftVoucherScreenWeb.class.getSimpleName();
+    private static final Logger LOGGER = Logger.getLogger(SCREEN_NAME);
+
+    private final By byPromoCodexpath = By.xpath(("//*[@id=\"PromoCode\"]"));
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/web/indigo/PromoScreenWeb.java:20]: bad xpath. can you tell me why?

File: src/test/java/com/znsio/sample/e2e/screen/web/indigo/LandingPageScreenWeb.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        this.innerDriver = this.driver.getInnerDriver();
+        long threadId = Thread.currentThread().getId();
+        context = Runner.getTestExecutionContext(threadId);
+    }
+    @Override
+    public GiftVoucherScreenWeb navigateToGiftVoucher(){
+
+        LOGGER.info("Navigate to Gift Voucher screen");
+        Actions action = new Actions(innerDriver);
+        action.moveToElement(driver.findElement(byBookDropdownXpath)).build().perform();
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/web/indigo/LandingPageScreenWeb.java:36]: indent properly

File: src/test/java/com/znsio/sample/e2e/screen/web/indigo/LandingPageScreenWeb.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+public class LandingPageScreenWeb extends LandingPageScreen {
+    private final Driver driver;
+    private final Visual visually;
+    private final WebDriver innerDriver;
+    private final TestExecutionContext context;
+    private static final String SCREEN_NAME = GiftVoucherScreenWeb.class.getSimpleName();
+    private static final Logger LOGGER = Logger.getLogger(SCREEN_NAME);
+
+    private final By byBookDropdownXpath = By.xpath("//a[@title='Book']");
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/web/indigo/LandingPageScreenWeb.java:21]: private static final

File: src/test/java/com/znsio/sample/e2e/screen/web/indigo/GiftVoucherScreenWeb.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    }
+        @Override
+         public boolean previewVoucher() {
+            LOGGER.info("Click Preview and Proceed ");
+            driver.scrollToBottom();
+            driver.waitTillElementIsPresent(byPreviewXpath).click();
+            visually.checkWindow(SCREEN_NAME,"Preview Voucher");
+            driver.findElement(byProceedXpath).click();
+            driver.waitTillElementIsPresent(byPromoCodexpath,10);
+            if(driver.isElementPresent(byPromoCodexpath)){
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/web/indigo/GiftVoucherScreenWeb.java:93]: this is bad. the screen method should be specific to each path.

File: src/test/java/com/znsio/sample/e2e/screen/web/indigo/GiftVoucherScreenWeb.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+        if(!makeItPersonalWebElement.isSelected())
+            makeItPersonalWebElement.click();
+        driver.waitTillElementIsPresent(byPersonalNameId).sendKeys((CharSequence) RecipientData.get(INDIGO_VOUCHER_CONTEXT.RECEIVER));
+
+        driver.waitTillElementIsPresent(byPersonalMessageId).sendKeys((CharSequence) RecipientData.get(INDIGO_VOUCHER_CONTEXT.MESSAGE));
+        visually.checkWindow(SCREEN_NAME,"Personalise Gift Voucher");
+        return this;
+    }
+        @Override
+         public boolean previewVoucher() {
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/web/indigo/GiftVoucherScreenWeb.java:86]: method name is incorrect. Does not indicate why a boolean is returned

File: src/test/java/com/znsio/sample/e2e/screen/web/indigo/GiftVoucherScreenWeb.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        return this;
+    }
+
+    @Override
+    public GiftVoucherScreen personaliseGiftVoucher(Map RecipientData) {
+
+        LOGGER.info("Personalise gift voucher with details "+ RecipientData.toString());
+        driver.scrollTillElementIntoView(byMakeItPersonalId);
+        WebElement makeItPersonalWebElement = driver.waitTillElementIsPresent(byMakeItPersonalId);
+        if(!makeItPersonalWebElement.isSelected())
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/web/indigo/GiftVoucherScreenWeb.java:77]: why is this if condition required?

File: src/test/java/com/znsio/sample/e2e/businessLayer/indigo/VoucherBL.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        LOGGER.info("Expected Voucher Amount : " + expectedTotalAmount);
+        data.put(INDIGO_VOUCHER_CONTEXT.EXPECTED_AMOUNT,String.valueOf(expectedTotalAmount));
+        String errorCode = promoScreen.applyPromoCode(data.get(INDIGO_VOUCHER_CONTEXT.PROMOCODE)).getErrorCode();
+        softly.assertThat(errorCode).contains("Invalid Promo Code");
+        return this;
+
+    }
+    public VoucherBL validatePurchase(){
+        int actualTotalAmount = PromoScreen.get().fetchActualPaymentAmount();
+        softly.assertThat(data.get(INDIGO_VOUCHER_CONTEXT.EXPECTED_AMOUNT)).isEqualTo(String.valueOf(actualTotalAmount));
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/businessLayer/indigo/VoucherBL.java:69]: assert description is missing

File: src/test/java/com/znsio/sample/e2e/businessLayer/indigo/VoucherBL.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+import java.util.Map;
+
+import static org.junit.Assert.assertThat;
+
+public class VoucherBL {
+    private static final Map<String, String> data = new LinkedHashMap<>();
+    private static final Logger LOGGER = Logger.getLogger(VoucherBL.class.getName());
+    public final TestExecutionContext context;
+    private final SoftAssertions softly;
+    Map<String, String> testData = Runner.getTestDataAsMap(System.getProperty("user.name"));
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/businessLayer/indigo/VoucherBL.java:24]: why do we have testData here?

File: src/test/java/com/znsio/sample/e2e/businessLayer/indigo/VoucherBL.java [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+import org.assertj.core.api.SoftAssertions;
+
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import static org.junit.Assert.assertThat;
+
+public class VoucherBL {
+    private static final Map<String, String> data = new LinkedHashMap<>();
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/businessLayer/indigo/VoucherBL.java:20]: why do we need this here?

File: src/test/java/com/znsio/sample/e2e/businessLayer/indigo/VoucherBL.java (hunk 13 above)
Comment (MSR-K) [reply src/test/java/com/znsio/sample/e2e/businessLayer/indigo/VoucherBL.java:24]: Unused Line , Forgot to remove. removing it.

File: src/test/java/com/znsio/sample/e2e/businessLayer/indigo/VoucherBL.java (hunk 14 above)
Comment (MSR-K) [reply src/test/java/com/znsio/sample/e2e/businessLayer/indigo/VoucherBL.java:20]: Setting Context value in Map  at Voucher BL Level.

**Pull request closed without merging**